croo/cromwell_metadata.py

```diff
@@ -137,31 +140,81 @@ def __parse_input_json(self):
+                Alias means a subworkflow imported with `as` statement in WDL.
+                `None` element in this list means a temporary subworkflow to implement
+                nested `scatter`.
+            parent_workflow_shard_indices:
+                Shard indices of all parent workflows (including nested scatter's indices).
+                Grander parent's index comes first.
+                The dimensions of `parent_workflows` and `parent_workflow_shard_indices` do not
+                necessarily match if there is a nested `scatter`.
         """
+        if not parent_workflows or not isinstance(parent_workflows, tuple):
+            raise ValueError(
+                'parent_workflows must be defined as non-empty tuple. '
+                'If it is a root calling of this function '
+                'then call with parent_workflows=(main_workflow_name,).'
+            )
+
         for call_name, call_list in calls.items():
-            for _, c in enumerate(call_list):
+            split_call_name = call_name.split('.')
+
+            if len(split_call_name) == 2:
+                subworkflow_or_task_alias = split_call_name[1]
+
+            elif len(split_call_name) == 1:
+                if not call_name.startswith('ScatterAt'):
+                    raise ValueError(
+                        'Wrong temporary call name format for a nested subworkflow.'
+                    )
+                subworkflow_or_task_alias = None
+
+            else:
+                raise ValueError('Wrong call name format. Too many dots.')
+
+            for c in call_list:
                 shard_idx = c['shardIndex']
-                if wf_alias is None:
-                    wf_name = call_name.split('.')[0]
-                else:
-                    wf_name = wf_alias
-                task_alias = call_name.split('.')[1]
 
                 # if it is a subworkflow, then recursively dive into it
                 if 'subWorkflowMetadata' in c:
                     self.__parse_calls(
                         c['subWorkflowMetadata']['calls'],
-                        parent_wf_name=parent_wf_name + wf_name + '.',
-                        wf_alias=task_alias,
-                        parent_wf_shard_idx=parent_wf_shard_idx + (shard_idx,),
+                        parent_workflows=parent_workflows
+                        + (subworkflow_or_task_alias,),
+                        parent_workflow_shard_indices=parent_workflow_shard_indices
+                        + (shard_idx,),
                     )
                     continue
 
-                task_name = parent_wf_name + wf_name + '.' + task_alias
-                shard_idx = parent_wf_shard_idx + (shard_idx,)
+                none_free_parent_workflows = parent_workflows + (
+                    subworkflow_or_task_alias,
+                )
+                none_free_parent_workflows = (
+                    wf for wf in none_free_parent_workflows if wf
+                )
```

Comment:
You could probably combine these into a single comprehension, saves reassigning.
```suggestion
                none_free_parent_workflows = (
                    workflow
                    for workflow
                    in parent_workflows + (subworkflow_or_task_alias,)
                    if workflow is not None
                )
```

Reply:
Fixed as you suggested.

Reply:
Yeah I think it's good time to add testing to Croo.
I will add basic linting and several integration tests to Croo.